Error types

zhengyao-lin opened the pull request:
Adding error types `ParseError` and `SerializeError` for debugging larger combinators.

This is WIP, but would appreciate some feedback.

TODO:
 - [x] Add an error combinator that always fails but produces a custom error message (e.g. `OrdChoice(OrdChoice(...), Error("Failed all choices"))`)
 - [ ] ~Record position of the error?~
 - [x] Change `vest-dsl` to work with right-associative OrdChoice

zhengyao-lin commented:
One issue with the error combinator is that it doesn't seem to work well with the OrdChoice being "left associative," due to how `DisjointFrom` is defined.

Operationally, we would want the Error combinator to be the last combinator to try
```
OrdChoice(
    OrdChoice(A, B),
    Error(...)
)
```
For this, we would need to implement `DisjointFrom` for `Error`:
```
impl<C: SpecCombinator> DisjointFrom<C> for Error { ... }
```

But this seem to conflict with `impl<U1, U2, U3> DisjointFrom<OrdChoice<U1, U2>> for U3`.
If we use the alternative "right-associative" version `impl<S1, S2, S3> DisjointFrom<S3> for OrdChoice<S1, S2>`, then the above trait impl would work.

y1ca1 commented:
After a first glance this looks good to me. I will take a closer look later this week.

I'd recommend naming the error combinator `Fail` or something similar.

It should be okay to make the impl of `Disjoint` for `OrdChoice` right associative without too much breaking, as I believe for the DSL all I need to do is to change a single place of the code where I used `fold` to `rfold` (hopefully).

y1ca1 commented:
> `impl<S1, S2, S3> DisjointFrom<S3> for OrdChoice<S1, S2>`

Actually, making this impl right-associative might be also beneficial to other combinators because I remember the left-associative `impl<U1, U2, U3> DisjointFrom<OrdChoice<U1, U2>> for U3` also creates conflict impl for e.g. `impl<U1, U2, U3> DisjointFrom<U3> for (U1, U2)`?

zhengyao-lin commented:
I changed the main `vest` library to use a right-associative OrdChoice, will work on `vest-examples` later.

I also added some macros to avoid specifying the associativity directly, e.g.
```
ord_choice!(a, b, c) = OrdChoice(a, OrdChoice(b, c))
ord_choice_type!(A, B, C) = OrdChoice<A, OrdChoice<B, C>>
ord_choice_result!(A, B, C) = Either<A, Either<B, C>>

inj_ord_choice_result!(m, *, *) = Either::Left(m)
inj_ord_choice_result!(*, m, *) = Either::Right(Either::Left(m))
inj_ord_choice_result!(*, *, m) = Either::Right(Either::Right(m))
```

The syntax of `inj_ord_choice_result` is not quite ideal. Let me know if prefer something else.

y1ca1 commented:
The DSL can generate those deeply nested `OrdChoice`s, `Either`s and stuff. But yeah, macros definitely makes sense for library users, we can have more of them maybe for `Mapped` and `pair` as well.

y1ca1 commented on vest/src/errors.rs:
```diff
@@ -0,0 +1,66 @@
+
+/// Sum of both parse and serialize errors
+#[derive(Debug)]
+pub enum Error {
```
Is this used anywhere?

zhengyao-lin replied:
It’s used in some tests with both parse and serialize, but nowhere else really.

y1ca1 commented on vest/src/regular/error.rs:
```diff
@@ -0,0 +1,69 @@
+/// Combinator used for custom error message
+pub struct CustomError(pub String);
+
+impl View for CustomError {
```
A bit nitpicking, personally I prefer the name "`Fail`" to indicate that this combinator would always fail. `CustomError`  per se doesn't sound like a combinator. If found useful, in future we might also introduce a "`Succeed`" combinator that always succeeds and produces no values.

zhengyao-lin replied:
Yeah Fail is a better name. I was just trying to avoid Error since it overlaps with the error type.

y1ca1 commented:
I'm hesitant about recording position of the error or even accumulating error context at this point, as it's known that error propagation in Rust _can_ be expensive, especially when it comes to deeply nested combinator chains. https://github.com/rust-bakery/nom/pull/1631 provides some insights as well.
